voting-dapp-template/move/sources/voting.move

```diff
@@ -0,0 +1,232 @@
+module voting_addr::voting {
+    use std::bcs;
+    use std::signer;
+    use std::string::String;
+    use std::vector;
+    use std::option;
+    use std::option::Option;
+    use aptos_framework::object::{Self, Object};
+    use aptos_framework::timestamp;
+    use aptos_std::string_utils;
+
+    // Error Codes
+    const E_PROPOSAL_DOES_NOT_EXIST: u64 = 1;
+    const E_USER_HAS_NO_GOVERNANCE_TOKENS: u64 = 2;
+    const E_USER_ALREADY_VOTED: u64 = 3;
+    const E_PROPOSAL_HAS_ENDED: u64 = 4;
+
+    // Global for contract
+    struct Proposal has key, store, drop, copy {
+        name: Option<String>,
+        creator: address,
+        start_time: u64,
+        end_time: u64,
+        yes_votes: u64,
+        no_votes: u64,
+    }
+
+    // Global for contract
+    struct ProposalRegistry has key, store, drop {
+        proposals: vector<Proposal>
+    }
+
+    // Unique for user
+    struct Vote has key, store {
+        voter: address,
+        vote: bool, // true for yes, false for no
+        amount: u64,
+    }
+
+    // This function is called once the module is published
+    fun init_module(sender: &signer) {
+        move_to(sender, ProposalRegistry {
+            proposals: vector::empty(),
+        });
+    }
+
+    // ======================== Write functions ========================
+    // function allows sender to create a new voting proposal
+    // There will be a start_time, end_time, and a name for the proposal
+    public entry fun create_proposal(sender: &signer, proposal_name: String, duration: u64) acquires ProposalRegistry {
+        let proposal_registry = borrow_global_mut<ProposalRegistry>(@voting_addr);
+        let curr = timestamp::now_seconds();
+        let new_proposal = Proposal {
+            name: option::some(proposal_name),
+            creator: signer::address_of(sender),
+            start_time: curr,
+            end_time: curr + duration,
+            yes_votes: 0,
+            no_votes: 0,
+        };
+        vector::push_back(&mut proposal_registry.proposals,new_proposal);
+    }
+
+    // write function that allows a user to vote on a proposal
+    public entry fun vote_on_proposal(sender: &signer, vote: bool, amount: u64) acquires ProposalRegistry {
+        // Check if a proposal exists
```

Comment:
i usually get the sender address here so i don't need to call `signer::address_of()` multiple times

```
let sender_addr = signer::address_of(sender);
```